### DISCUSSION TITLE
Cannot build due to com.apple.developer.web-browser entitlement

### DISCUSSION THREAD

**Discussion by fuomag9** [blinksh/blink#1746] [Support] [answered]
### Checklist

- [x] Using latest version on the [App Store](http://itunes.apple.com/app/id1156707581)
- [x] Read the [docs](http://blink.sh/docs/)
- [x] Searched for [existing GitHub issues](https://github.com/blinksh/blink/issues)

### Configuration

None relevant

### Describe the bug

I cannot build due to 

```
Provisioning profile "iOS Team Provisioning Profile: com.examplferfver4gfrgte.app" doesn't include the com.apple.developer.web-browser entitlement.
```

I should be able to disable the need to use the browser entitlement

![Screenshot 2023-03-25 at 11 49 14](https://user-images.githubusercontent.com/1580624/227712729-35952616-a23a-4440-bdb4-8bdf865d7fc4.png)

**Comment by smorkin** [comment]
Is this entitlement needed? According to Apple docs, https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_developer_web-browser, this is for apps which can act as the default web browser on iOS, so e.g. Mozilla mobile uses it (see this bug, https://github.com/mozilla-mobile/focus-ios/issues/1781, where they deal with the same issue).

**Comment by carloscabanero** [comment, marked as answer]
This entitlement is needed for Blink Code. You can remove it and compile without it. Just take into account that you will have to use a tunnel if you use Blink Code with remote servers.

**Reply by recipedude** [reply]
Rather than remove the entitlement, how does one resolve the build error reported (getting same error here) by enabling the entitlement?

**Reply by recipedude** [reply]
Following up on this, I created a non-Xcode managed profile.  In the Edit your App ID Configuration section I ticked off every Capability I possibly could.  I then created a provisioning profile using those the "everything" Identifier, downloaded and selected that provisioning profile in Xcode.  

And still get the error:
`Provisioning profile "BlinkMy IOS with Everything" doesn't include the com.apple.developer.system-extension.install and com.apple.developer.web-browser entitlements.` 

How the heck does one get those required entitlements added?

**Reply by carloscabanero** [reply]
We had to get a special permission from Apple to actually enable those entitlements and get them to work.

**Reply by recipedude** [reply]
Ah, lovely - well I did submit a bug report to Apple about an hour ago on this due to the failure of the “Xcode provisioning to provision” :).  Have a feeling the response won’t be a fast as yours though! 👍

**Comment by tedserreyn** [comment]
Ok so how does one actually remove the WEB provisioning to compile this?

**Reply by romapires** [reply]
Search for _com.apple.developer.web-browser_ in the xcode.
It's a Boolen key in Blink/Blink.entitlements  set to YES (or true in plain text).

The key must be removed, it's not enough to invert it.

**Comment by carloscabanero** [comment]
Does anybody know if we can wrap this under a flag? So it doesn’t have to be removed every time?

**Reply by romapires** [reply]
I compiled Blink yesterday with a personal team. Great app btw!

I found out many Capabilities don't work with a personal team, namely the services with these keys in the Entitlements file:

- aps-environment
- com.apple.developer.associated-domains
- com.apple.developer.icloud-container-identifiers
- com.apple.developer.icloud-services
- com.apple.developer.ubiquity-container-identifiers
- com.apple.developer.ubiquity-kvstore-identifier
- com.apple.developer.user-fonts
- com.apple.developer.web-browser

The fix I see, and can contribute with by tomorrow, is:

1. Add a `PERSONAL_TEAM = FALSE` to _template_setup.xcconfig_
2. Add a Build Script named something like "Personal Team Entitlements fix" that 
    1. Checks the env for PERSONAL_TEAM = TRUE
    2. Checks if **PlistBuddy** exists (its from command line tools but… you never know)
    3. Uses **PlistBuddy** to delete the offending keys from _Blink.entitlements_

Going forward, the usages of the these Capabilities could be wrapped with `#if CAN_USE_capabilityName … #endif` defined by negating PERSONAL_TEAM.

Does that look ok to you?

**Reply by carloscabanero** [reply]
That sounds good from my side. I like that the default for the script is to do nothing (even not running) and then we can define flags going forward, or even separate on different files if necessary. 

Thanks a lot!

**Reply by yihongang** [reply]
Hi, I noticed the same thing @romapires did (latest `raw` code does not build for personal teams due to certain capabilities being blocked for such teams) and also that neither the `PERSONAL_TEAM` env  nor the CAN_USE directives exist in the code. Is that still a welcomed change?

**Reply by carloscabanero** [reply]
Yeah, fixes for this are more than welcome.

**Reply by maelp** [reply]
Is there now an option to do this?

**Reply by carloscabanero** [reply]
I think just removing the entitlements that don't work for you should do the trick. As for the Web-Browser entitlement, which is the tricky one, I'm thinking about separating Blink Code in its own app, as things are growing to a point where it doesn't make sense to have everything stitched together.

If you end up creating the flag for the entitlements, which hopefully shouldn't be hard, please feel free to contribute it and I will merge it from my side. I unfortunately have no idea how to tweak and bend Xcode that way.
